src/framework/annotator_mgr.cpp

```diff
@@ -583,7 +576,185 @@ namespace uima {
+                std::unique_ptr<Flow> nextFlow = flow->newCasProduced(*outputCas, nextAEKey);
+                casIterStack.push({nextAE, currentCas, std::move(flow), nextAEKey});
+                flow = std::move(nextFlow);
+                currentCas = outputCas;
+                activeCASes.insert(currentCas);
+              } else {
+                // No new CASes are output, this CAS is done being processed by the current engine.
+                currentCas->setCurrentComponentInfo(nullptr);
+              }
+            } else {
+              UIMA_EXC_THROW_NEW(EngineProcessingException,
+                                 UIMA_ERR_USER_ANNOTATOR_COULD_NOT_PROCESS,
+                                 UIMA_MSG_ID_EXCON_PROCESSING_CAS,
+                                 ErrorMessage(UIMA_MSG_ID_LITERAL_STRING, "Unknown Delegate Key"),
+                                 ErrorInfo::unrecoverable);
+            }
+          } else if (nextStep.getType() == Step::StepType::PARALLELSTEP) {
+            // TODO: ParallelStep not supported yet
+            UIMA_EXC_THROW_NEW(NotYetImplementedException,
+                               UIMA_ERR_NOT_YET_IMPLEMENTED,
+                               UIMA_MSG_ID_EXC_NOT_YET_IMPLEMENTED,
+                               ErrorMessage(UIMA_MSG_ID_LITERAL_STRING, "Parallel Step not supported yet"),
+                               ErrorInfo::unrecoverable
+            );
+          } else {
+            UIMA_EXC_THROW_NEW(EngineProcessingException,
+                               UIMA_ERR_USER_ANNOTATOR_COULD_NOT_PROCESS,
+                               UIMA_MSG_ID_EXCON_PROCESSING_CAS,
+                               ErrorMessage(UIMA_MSG_ID_LITERAL_STRING, "Unknown Step Type"),
+                               ErrorInfo::unrecoverable);
+          }
+
+          nextStep = flow->next();
+        }
+
+        const FinalStep *finalStep = nextStep.getFinalStep();
+        if (currentCas == inputCas) {
+          if (finalStep->getForceDropCAS()) {
+            // Not allowed to drop the input CAS so something must have gone wrong
+            UIMA_EXC_THROW_NEW(EngineProcessingException,
+                               UIMA_ERR_USER_ANNOTATOR_COULD_NOT_PROCESS,
+                               UIMA_MSG_ID_EXCON_PROCESSING_CAS,
+                               ErrorMessage(UIMA_MSG_ID_LITERAL_STRING, "Illegal CAS drop"),
+                               ErrorInfo::unrecoverable);
+          }
+          return nullptr;
+        }
+
+        if (iv_bOutputNewCases && !finalStep->getForceDropCAS())
+          return currentCas;
+        currentCas->release();
+      }
+    }
+
+    bool AnnotatorManager::hasNext() {
+      if (!nextCas)
+        nextCas = processUntilNextOutputCas();
+      return nextCas != nullptr;
+    }
+
+    CAS & AnnotatorManager::next() {
+      CAS* result = nextCas;
+      if (!result)
+        result = processUntilNextOutputCas();
+      if (!result) {
+        UIMA_EXC_THROW_NEW(Exception,
```

Comment:
shouldn't this call also `release()`?

Reply:
No when it is returned to the caller/user they will call `release` themselves. Or do you mean something else?